src/ui/GameUI.cpp

```diff
@@ -0,0 +1,358 @@
+// GameUI.cpp
+
+#include "GameUI.hpp"
+#include "Frame.hpp"
+#include "Image.hpp"
+#include "Field.hpp"
+
+#include "../main.hpp"
+#include "../game.hpp"
+#include "../menu.hpp"
+#include "../interface/interface.hpp"
+
+#include <assert.h>
+
+static Frame* playerHud[MAXPLAYERS] = { nullptr };
+
+void createIngameHud(int player) {
+    char name[32];
+    snprintf(name, sizeof(name), "player hud %d", player);
+    Frame* frame = gui->addFrame(name);
+
+    playerHud[player] = frame;
+    int playercount = 0;
+    if (multiplayer == SINGLE) {
+        for (int c = 0; c < MAXPLAYERS; ++c) {
+            if (!client_disconnected[c]) {
+                ++playercount;
+            }
+        }
+    } else {
+        playercount = 1;
+    }
+
+    static const char* bigfont = "fonts/pixelmix.ttf#18";
+    static const char* smallfont = "fonts/pixel_maz.ttf#32";
+
+    // big empty frame to serve as the root
+    if (playercount == 1) {
+        frame->setSize(SDL_Rect{0, 0, Frame::virtualScreenX, Frame::virtualScreenY});
+    } else if (playercount == 2) {
+        int y = (player % 2) * Frame::virtualScreenY / 2;
+        frame->setSize(SDL_Rect{0, y, Frame::virtualScreenX, Frame::virtualScreenY / 2});
+    } else if (playercount >= 3 || playercount <= 4) {
```

Comment:
What about the 16 player beta branch? :)

Reply:
I assume that playing a multiplayer game with 16 network but 1 local player works just fine here, right?

Reply:
Yeah multiplayer != SINGLE covers any server multiplayer configuration.

FWIW I've got the following helper functions I've been using to count players inside the player[] array
```
const bool Player::isLocalPlayer() const
{
	return ((splitscreen && bSplitscreen) || playernum == clientnum);
}
const bool Player::isLocalPlayerAlive() const
{
	return (isLocalPlayer() && entity && !client_disconnected[playernum]);
}
```